enhancements/network/sdn-live-migration.md

```diff
@@ -0,0 +1,358 @@
+   “migration.network.openshift.io/plugin: OVNKubernetes”). Since there are not
+   currently any nodes labeled that way, this means no actual ovnkube-node pods
+   will be created at this step.
+
+At this point, neither plugin is willing to run on a node that has no
+“migration.network.openshift.io/plugin” label at all. This means that if a new
+node gets added to the cluster in the middle of the migration process, neither
+plugin will get deployed to it until after the migration of the older nodes is
+complete, and the node affinity is removed.
+
+#### Migration
+
+In order to run ovn-kubernetess in shared-gw mode, we need MCO to update the
+ovs-configuration.service. To apply the MachineConfig update, a reboot will be
+triggered by MCO. To coordinate the draining and rebooting between CNO and MCO,
+we need to modify MCO as following:
+- add a `machineconfiguration.openshift.io/rebootPausedBy: network-operator`
+  node annotation to MCO, so that CNO can block MCO from rebooting the node
+- add a new value `RebootPending` to the existing
+  `machineconfiguration.openshift.io/state` node annotation. It will be set by
+  MCD if the reboot is blocked. It indicts that draining node is done, and the
+  rebooting is paused.
+
+When we kick off the migration from CNO, CNO will update the
+`status.networkType` field of the `network.config` CR. It will trigger MCO to
+apply new MachineConfig to each node:
+1. MCO will 
+   - rerender the MachineConfig for each node
+   - try to cordon and drain the node one node
+2. CNO finds a node is being drained by MCO by watching the MCO node
+   annotations, it will:
+   - pause the node reboot by annotate the node with
+     `machineconfiguration.openshift.io/rebootPausedBy: network-operator`.
+3. MCO will set node annotation `machineconfiguration.openshift.io/state:
+   RebootPending`, when draining node is completed.
+4. CNO finds the node has been drained by MCO, it will:
+   - annotate the node with ovnkube annotations:
+     `k8s.ovn.org/node-subnets: '{"default":"10.131.0.0/23"}'`
+   - remove the "migration.network.openshift.io/plugin" label from the node wait
+     for the sdn-node DaemonSet to decrement its NumberReady
+   - label the node "migration.network.openshift.io/plugin: OVNKubernetes"
+   - remove the `rebootPausedBy` node annotation.
+5. MCO will reboot the node
+6. After booting up, br-ex will be created and ovnkube-node will run in hybrid
+   overlay mode on the node.
+7. MCO will uncordon the node. Pods will be created on the node using
+   ovn-kubernetes as the default CNI plugin.
+
+The above process will be repeated for each node, until all the nodes have been
+applied to the new MachineConfig and converted to OVN Kubernetes.
+
+#### Migration Cleanup
+
+Once migration is complete, CNO will:
+- delete the openshift-sdn DaemonSets
+- redeploy ovn-kubernetes in “normal” mode (no migration mode config, no node
+  affinity).
```

Comment:
Nodes are uncordoned at this point, so OVN is redeployed with workloads deployed to the node.  Is it possible to switch this over node-by-node in the migration steps above to reduce workload impacts?

Reply:
This step is about fixing the state of the `ovnkube-master` and `ovnkube-node` Deployment objects. While the Deployment controller will roll the update out node-by-node, there's nothing we can do specially, since we're just talking about about updating two API objects.

Also note that all of the nodes are already running ovn-kubernetes at this point, so this is just a question of restarting the existing ovn-kubernetes binaries on each node. It should be very quick/non-disruptive.